ai_oca_bridge_extra_parameters/models/ai_extra_parameter.py

```diff
@@ -0,0 +1,510 @@
+            ("formula", "Formula"),
+        ],
+        required=True,
+        default="expression",
+        help="Defines how the parameter is evaluated. "
+        "'expression' means it will be evaluated as a Python expression, "
+        "'formula' means it will be evaluated as a formula expression.",
+    )
+    FORBIDDEN_PATTERNS = [
+        r"\.cr\b",
+        r"\._cr\b",
+        r"\.execute\s*\(",
+        r"\.commit\s*\(",
+        r"\.rollback\s*\(",
+        r"\.sudo\s*\(",
+        r"\.with_context\s*\(",
+        r"\.write\s*\(",
+        r"\.unlink\s*\(",
+        r"\.invalidate_cache\s*\(",
+        r"\.flush\s*\(",
+        r"\.flush_all\s*\(",
+        r"\bimport\b",
+        r"\b(exec|eval)\s*\(",
+        r"\bopen\s*\(",
+        r"\bgetattr\s*\(",
+        r"__\w+__",
+    ]
+
+    @staticmethod
+    def _extract_variables(expr):
+        try:
+            tree = ast.parse(expr, mode="eval")
+            extractor = ExpressionVariableExtractor()
+            extractor.visit(tree)
+            return extractor.variables
+        except Exception:
+            return set()
+
+    @staticmethod
+    def _extract_variables_from_code(code):
+        try:
+            tree = ast.parse(code, mode="exec")
+            extractor = FormulaVariableExtractor()
+            extractor.visit(tree)
+            return extractor.external_vars
+        except Exception:
+            return set()
+
+    @classmethod
+    def _validate_expression(cls, expr):
+        try:
+            ast.parse(expr, mode="eval")
+        except SyntaxError as e:
+            return False, f"Invalid Python syntax: {e}"
+        except Exception as e:
+            return False, f"Invalid expression: {e}"
+
+        for pattern in cls.FORBIDDEN_PATTERNS:
+            if re.search(pattern, expr):
+                if r"\(" in pattern:
+                    op_match = re.search(r"\\\.([\w_]+)", pattern)
+                else:
+                    op_match = re.search(r"\\\.([\w_]+)", pattern)
+                operation = op_match.group(1) if op_match else "restricted operation"
+                return False, f"Access to '{operation}' is not allowed"
+
+        variables = cls._extract_variables(expr)
+        allowed_names = {
+            "object",
+            "True",
+            "False",
+            "None",
+            "str",
+            "int",
+            "float",
+            "bool",
+            "len",
+            "list",
+            "map",
+            "filter",
+            "zip",
+            "sum",
+            "min",
+            "max",
+            "sorted",
+            "enumerate",
+            "range",
+        }
+        invalid_vars = variables - allowed_names
+
+        if invalid_vars:
+            return (
+                False,
+                f"Only 'object' variable is allowed, found: {', '.join(invalid_vars)}",
+            )
+
+        return True, ""
+
+    @classmethod
+    def _validate_formula(cls, formula):
```

Comment:
the methods: "_validate_formula" & "_validate_expression" are more similars, can u join them?